core/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -189,8 +190,11 @@ public RepositoryData addSnapshot(final SnapshotId snapshotId,
      */
     public RepositoryData removeSnapshot(final SnapshotId snapshotId) {
         Map<String, SnapshotId> newSnapshotIds = snapshotIds.values().stream()
-            .filter(id -> snapshotId.equals(id) == false)
+            .filter(id -> !snapshotId.equals(id))
             .collect(Collectors.toMap(SnapshotId::getUUID, Function.identity()));
+        if (newSnapshotIds.size() == snapshotIds.size()) {
+            throw new ElasticsearchException("Attempting to remove non-existent snapshot [{}] from repository data", snapshotId);
```

Comment:
Perhaps, something like `ResourceNotFoundException` would be more appropriate here?